### DISCUSSION TITLE
EIP-712 signing fails on eth_signTypedData setup

### DISCUSSION THREAD

**Discussion by onkaraman** [trustwallet/wallet-core#3049] [Q&A]
### Discussed in https://github.com/trustwallet/wallet-core/discussions/3045

<div type='discussions-op-text'>

<sup>Originally posted by **onkaraman** March 31, 2023</sup>
Hi all,

the following code works with MetaMask, but on Trust Wallet I have issues:

```
let signature = await window.ethereum.request({
	method: 'eth_signTypedData_v4',
	params: [connectedAddress,
		JSON.stringify(EIP712types.newRecord(this.state.createName, this.state.createSNr))
	]
});
```

The value of `JSON.stringify(EIP712types.newRecord(this.state.createName, this.state.createSNr))` is
```
{
	"types": {
		"EIP712Domain": [{
			"name": "name",
			"type": "string"
		}, {
			"name": "version",
			"type": "string"
		}, {
			"name": "chainId",
			"type": "uint256"
		}, {
			"name": "verifyingContract",
			"type": "address"
		}],
		"NewRecord": [{
			"name": "productName",
			"type": "string"
		}, {
			"name": "productUID",
			"type": "string"
		}]
	},
	"primaryType": "NewRecord",
	"domain": {
		"name": "MyDomain",
		"version": "1",
		"chainId": 1337,
		"verifyingContract": "..."
	},
	"message": {
		"productName": "MGS1PS1",
		"productUID": "MGS1PS1"
	}
}
```

Trust Wallet returns the error:

```
SyntaxError: Unexpected non-whitespace character after JSON at position 1
    at JSON.parse (<anonymous>)
    at g.value (inpage.js:2:42406)
    at inpage.js:2:40108
```

I'm testing my dApp locally with ganache.

</div>

**Comment by zlace0x** [comment]
I'm facing a similar issue on `eth_signTypedData` only on trust iOS app (works on android) via walletconnect, while signing a erc20 permit works fine.

```javascript
// Domain 
{
    types: {
      MetaTransaction: [
        { name: "nonce", type: "uint256" },
        { name: "from", type: "address" },
        { name: "functionSignature", type: "bytes" },
      ],
    },
    primaryType: "MetaTransaction",
    domain: {
      name: domainName,
      version: "1",
      chainId,
      verifyingContract,
    },
    message: {
      nonce,
      from,
      functionSignature,
    },
  }
}
// response
{
    "code": -32000,
    "message": "The operation couldn’t be completed. (TrustWalletConnect.WCError error 1.)"
}
```

**Reply by Doberman2029** [reply]
Same problem. Trust wallet app on ios + desktop wallet connect qr code. But only with accounts created through trust wallet

**Comment by xieweizhi** [comment]
Hi @onkaraman , to help us better locate the issue, please send us the raw message that is sent to Trust.

**Reply by alexshchur** [reply]
Hi @xieweizhi , I'm also facing problems when trying to sign typed data
I can consistently reproduce the error in this sandbox https://codesandbox.io/s/white-rain-8f9pq9

And here's a snippet with the exact message:
```
  const args = {
    method: "eth_signTypedData_v4",
    params: [
      accounts[0],
      '{"types":{"Order":[{"name":"nonceAndMeta","type":"uint256"},{"name":"expiry","type":"uint128"},{"name":"makerAsset","type":"address"},{"name":"takerAsset","type":"address"},{"name":"maker","type":"address"},{"name":"taker","type":"address"},{"name":"makerAmount","type":"uint256"},{"name":"takerAmount","type":"uint256"}],"EIP712Domain":[{"name":"name","type":"string"},{"name":"version","type":"string"},{"name":"chainId","type":"uint256"},{"name":"verifyingContract","type":"address"}]},"domain":{"name":"AUGUSTUS RFQ","version":"1","chainId":"56","verifyingContract":"0x8dcdfe88ef0351f27437284d0710cd65b20288bb"},"primaryType":"Order","message":{"nonceAndMeta":"3100080921667902682720200198275800054486619795385893978032308224","expiry":"0","makerAsset":"0x1af3f329e8be154074d8769d1ffa4ee058b1dbc3","takerAsset":"0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c","maker":"0xeb151d9d151843e990d33a4e575454a0e8e1068f","taker":"0x0000000000000000000000000000000000000000","makerAmount":"1000000000000000","takerAmount":"4300000000000"}}'
    ]
  };
  const signed = await provider.request(args);
```

**Reply by xieweizhi** [reply]
Hi, @alexshchur , I think you need to `JSON.stringify` the message.

After stringify the message:

![IMG_005E45CF3F9A-1](https://github.com/trustwallet/wallet-core/assets/1779863/56e88bce-0345-4818-bc50-2550403a1f04)

We are considering to support unstringify message. But before that, please stringify your message.

**Reply by alexshchur** [reply]
Hey @onkaraman , thanks for your reply.
I do stringify the object. What I'm seeing on your screenshot looks like a standard `personal_sign` call, whilst what I need is `eth_signTypedData_v4`.

I also opened your own sandbox the URL of which I noticed on the screenshot (https://codesandbox.io/s/9xdpfs) and I can see the same problem is reproduced there

Looks like `eth_signTypedData_v4` Trustwallet IOS build is broken for the Multi-Coin wallets, in a case when the current chain in the trust wallet is not the mainnet. The same bug is observed on Uniswap

**Reply by alexshchur** [reply]
I believe it's a bug, so I created this [bug report](https://github.com/trustwallet/wallet-core/issues/3304)
